<a href="https://colab.research.google.com/github/jstadden34/jstadden_adv_mach_learn/blob/master/jstadden_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Original Example:

In [11]:
import keras
keras.__version__

from keras.datasets import imdb
(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words=10000)

train_data[0]

train_labels[0]

max([max(sequence) for sequence in train_data])



import numpy as np

def vectorize_sequences(sequences, dimension=10000):
    # Create an all-zero matrix of shape (len(sequences), dimension)
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.  # set specific indices of results[i] to 1s
    return results


# Our vectorized training data
x_train = vectorize_sequences(train_data)
# Our vectorized test data
x_test = vectorize_sequences(test_data)

x_train[0]


# Our vectorized labels
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')



from keras import models
from keras import layers

model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


from keras import optimizers

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


from keras import losses
from keras import metrics

model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

<string>:6: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/datasets/imdb.py:159: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/datasets/imdb.py:160: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 

Epoch 1/20
30/30 [==============================] - 2s 45ms/step - loss: 0.6123 - binary_accuracy: 0.6845 - val_loss: 0.4235 - val_binary_accuracy: 0.8606
Epoch 2/20
30/30 [==============================] - 1s 35ms/step - loss: 0.3618 - binary_accuracy: 0.8933 - val_loss: 0.3234 - val_binary_accuracy: 0.8856
Epoch 3/20
30/30 [==============================] - 1s 36ms/step - loss: 0.2477 - binary_accuracy: 0.9321 - val_loss: 0.2846 - val_binary_accuracy: 0.8928
Epoch 4/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1941 - binary_accuracy: 0.9424 - val_loss: 0.2761 - val_binary_accuracy: 0.8922
Epoch 5/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1517 - binary_accuracy: 0.9543 - val_loss: 0.2774 - val_binary_accuracy: 0.8903
Epoch 6/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1198 - binary_accuracy: 0.9661 - val_loss: 0.2915 - val_binary_accuracy: 0.8859
Epoch 7/20
30/30 [==============================] - 1s 35ms/step - los

[0.2960135340690613, 0.8830400109291077]

Comment: Accuracy = 88.3%

1 a) Change: 1 hidden layer

In [12]:

(train_data, train_labels), (test_data, test_labels) = imdb.load_data(num_words=10000)

train_data[0]

train_labels[0]

max([max(sequence) for sequence in train_data])


def vectorize_sequences(sequences, dimension=10000):
    # Create an all-zero matrix of shape (len(sequences), dimension)
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        results[i, sequence] = 1.  # set specific indices of results[i] to 1s
    return results

# Our vectorized training data
x_train = vectorize_sequences(train_data)
# Our vectorized test data
x_test = vectorize_sequences(test_data)



# Our vectorized labels
y_train = np.asarray(train_labels).astype('float32')
y_test = np.asarray(test_labels).astype('float32')



model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

<string>:6: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/datasets/imdb.py:159: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/datasets/imdb.py:160: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 

Epoch 1/20
30/30 [==============================] - 2s 44ms/step - loss: 0.5681 - binary_accuracy: 0.7296 - val_loss: 0.3904 - val_binary_accuracy: 0.8711
Epoch 2/20
30/30 [==============================] - 1s 35ms/step - loss: 0.3332 - binary_accuracy: 0.9027 - val_loss: 0.3203 - val_binary_accuracy: 0.8861
Epoch 3/20
30/30 [==============================] - 1s 35ms/step - loss: 0.2577 - binary_accuracy: 0.9232 - val_loss: 0.3056 - val_binary_accuracy: 0.8808
Epoch 4/20
30/30 [==============================] - 1s 35ms/step - loss: 0.2122 - binary_accuracy: 0.9348 - val_loss: 0.2814 - val_binary_accuracy: 0.8907
Epoch 5/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1716 - binary_accuracy: 0.9501 - val_loss: 0.2753 - val_binary_accuracy: 0.8903
Epoch 6/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1492 - binary_accuracy: 0.9562 - val_loss: 0.2732 - val_binary_accuracy: 0.8913
Epoch 7/20
30/30 [==============================] - 1s 34ms/step - los

[0.3003613352775574, 0.8784000277519226]

Discuss: Reducing the number of hidden layers led to a slightly lower accuracy of 87.8%

1 b) Change: 3 hidden layers

In [13]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

Epoch 1/20
30/30 [==============================] - 2s 45ms/step - loss: 0.6069 - binary_accuracy: 0.6844 - val_loss: 0.3789 - val_binary_accuracy: 0.8767
Epoch 2/20
30/30 [==============================] - 1s 35ms/step - loss: 0.3141 - binary_accuracy: 0.9011 - val_loss: 0.2968 - val_binary_accuracy: 0.8859
Epoch 3/20
30/30 [==============================] - 1s 36ms/step - loss: 0.2075 - binary_accuracy: 0.9362 - val_loss: 0.3555 - val_binary_accuracy: 0.8545
Epoch 4/20
30/30 [==============================] - 1s 36ms/step - loss: 0.1677 - binary_accuracy: 0.9444 - val_loss: 0.3436 - val_binary_accuracy: 0.8621
Epoch 5/20
30/30 [==============================] - 1s 36ms/step - loss: 0.1353 - binary_accuracy: 0.9557 - val_loss: 0.3368 - val_binary_accuracy: 0.8688
Epoch 6/20
30/30 [==============================] - 1s 36ms/step - loss: 0.1011 - binary_accuracy: 0.9720 - val_loss: 0.3175 - val_binary_accuracy: 0.8810
Epoch 7/20
30/30 [==============================] - 1s 35ms/step - los

[0.31819429993629456, 0.8772000074386597]

Discuss: Increasing to 3 hidden layers also resulted in a slightly lower accuracy of 87.7%

2 a) Change: 32 hidden units

In [14]:
model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(32, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(32, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

Epoch 1/20
30/30 [==============================] - 2s 56ms/step - loss: 0.5589 - binary_accuracy: 0.7151 - val_loss: 0.3251 - val_binary_accuracy: 0.8823
Epoch 2/20
30/30 [==============================] - 1s 42ms/step - loss: 0.2772 - binary_accuracy: 0.9063 - val_loss: 0.2781 - val_binary_accuracy: 0.8916
Epoch 3/20
30/30 [==============================] - 1s 42ms/step - loss: 0.1850 - binary_accuracy: 0.9392 - val_loss: 0.3127 - val_binary_accuracy: 0.8728
Epoch 4/20
30/30 [==============================] - 1s 42ms/step - loss: 0.1561 - binary_accuracy: 0.9460 - val_loss: 0.2835 - val_binary_accuracy: 0.8891
Epoch 5/20
30/30 [==============================] - 1s 42ms/step - loss: 0.1095 - binary_accuracy: 0.9649 - val_loss: 0.3027 - val_binary_accuracy: 0.8834
Epoch 6/20
30/30 [==============================] - 1s 42ms/step - loss: 0.0936 - binary_accuracy: 0.9691 - val_loss: 0.3328 - val_binary_accuracy: 0.8772
Epoch 7/20
30/30 [==============================] - 1s 42ms/step - los

[0.3092631995677948, 0.8803200125694275]

Discuss: Increasing to 32 hidden units led to a slightly lower accuracy of 88.0%

2 b) Change: 8 hidden units

In [15]:
model = models.Sequential()
model.add(layers.Dense(8, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(8, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(8, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(8, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

Epoch 1/20
30/30 [==============================] - 2s 57ms/step - loss: 0.6037 - binary_accuracy: 0.7178 - val_loss: 0.4274 - val_binary_accuracy: 0.8585
Epoch 2/20
30/30 [==============================] - 1s 35ms/step - loss: 0.3685 - binary_accuracy: 0.8898 - val_loss: 0.3341 - val_binary_accuracy: 0.8775
Epoch 3/20
30/30 [==============================] - 1s 35ms/step - loss: 0.2613 - binary_accuracy: 0.9211 - val_loss: 0.2911 - val_binary_accuracy: 0.8905
Epoch 4/20
30/30 [==============================] - 1s 35ms/step - loss: 0.2069 - binary_accuracy: 0.9370 - val_loss: 0.2770 - val_binary_accuracy: 0.8911
Epoch 5/20
30/30 [==============================] - 1s 36ms/step - loss: 0.1744 - binary_accuracy: 0.9457 - val_loss: 0.2725 - val_binary_accuracy: 0.8911
Epoch 6/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1415 - binary_accuracy: 0.9588 - val_loss: 0.2865 - val_binary_accuracy: 0.8857
Epoch 7/20
30/30 [==============================] - 1s 36ms/step - los

[0.2818399667739868, 0.8884000182151794]

Discuss: Reducing to 8 hidden units gave a slightly higher accuracy of 88.8%

3 Change: mse loss function

In [16]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='mse',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='mse',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.mse,
              metrics=[metrics.mse])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='mse',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

Epoch 1/20
30/30 [==============================] - 2s 45ms/step - loss: 0.2006 - mean_squared_error: 0.2006 - val_loss: 0.1217 - val_mean_squared_error: 0.1217
Epoch 2/20
30/30 [==============================] - 1s 35ms/step - loss: 0.0931 - mean_squared_error: 0.0931 - val_loss: 0.0969 - val_mean_squared_error: 0.0969
Epoch 3/20
30/30 [==============================] - 1s 35ms/step - loss: 0.0655 - mean_squared_error: 0.0655 - val_loss: 0.0840 - val_mean_squared_error: 0.0840
Epoch 4/20
30/30 [==============================] - 1s 35ms/step - loss: 0.0471 - mean_squared_error: 0.0471 - val_loss: 0.0859 - val_mean_squared_error: 0.0859
Epoch 5/20
30/30 [==============================] - 1s 35ms/step - loss: 0.0357 - mean_squared_error: 0.0357 - val_loss: 0.0827 - val_mean_squared_error: 0.0827
Epoch 6/20
30/30 [==============================] - 1s 35ms/step - loss: 0.0283 - mean_squared_error: 0.0283 - val_loss: 0.0896 - val_mean_squared_error: 0.0896
Epoch 7/20
30/30 [================

[0.08649957925081253, 0.8823599815368652]

Discuss: Using the mse loss function gave a nearly equivalent accuracy of 88.2%

4 Change: tanh activation

In [17]:
model = models.Sequential()
model.add(layers.Dense(16, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(16, activation='tanh'))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(16, activation='tanh', input_shape=(10000,)))
model.add(layers.Dense(16, activation='tanh'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

Epoch 1/20
30/30 [==============================] - 2s 45ms/step - loss: 0.5738 - binary_accuracy: 0.7165 - val_loss: 0.3593 - val_binary_accuracy: 0.8759
Epoch 2/20
30/30 [==============================] - 1s 35ms/step - loss: 0.2927 - binary_accuracy: 0.9096 - val_loss: 0.2872 - val_binary_accuracy: 0.8912
Epoch 3/20
30/30 [==============================] - 1s 35ms/step - loss: 0.2005 - binary_accuracy: 0.9362 - val_loss: 0.2902 - val_binary_accuracy: 0.8828
Epoch 4/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1459 - binary_accuracy: 0.9530 - val_loss: 0.2910 - val_binary_accuracy: 0.8827
Epoch 5/20
30/30 [==============================] - 1s 35ms/step - loss: 0.1080 - binary_accuracy: 0.9678 - val_loss: 0.3040 - val_binary_accuracy: 0.8837
Epoch 6/20
30/30 [==============================] - 1s 35ms/step - loss: 0.0800 - binary_accuracy: 0.9766 - val_loss: 0.3727 - val_binary_accuracy: 0.8740
Epoch 7/20
30/30 [==============================] - 1s 35ms/step - los

[0.4177379608154297, 0.8479200005531311]

Discussion: Using tanh activation led to a lower accuracy of 84.8%

5 a) Change: Regularization

In [18]:
from keras import regularizers

model = models.Sequential()
model.add(layers.Dense(16, kernel_regularizer=regularizers.l2(0.001), activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, kernel_regularizer=regularizers.l2(0.001), activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

Epoch 1/20
30/30 [==============================] - 2s 46ms/step - loss: 0.6339 - binary_accuracy: 0.6994 - val_loss: 0.4445 - val_binary_accuracy: 0.8592
Epoch 2/20
30/30 [==============================] - 1s 36ms/step - loss: 0.3828 - binary_accuracy: 0.8994 - val_loss: 0.3641 - val_binary_accuracy: 0.8824
Epoch 3/20
30/30 [==============================] - 1s 35ms/step - loss: 0.2934 - binary_accuracy: 0.9233 - val_loss: 0.3330 - val_binary_accuracy: 0.8906
Epoch 4/20
30/30 [==============================] - 1s 36ms/step - loss: 0.2503 - binary_accuracy: 0.9380 - val_loss: 0.3375 - val_binary_accuracy: 0.8851
Epoch 5/20
30/30 [==============================] - 1s 36ms/step - loss: 0.2306 - binary_accuracy: 0.9437 - val_loss: 0.3326 - val_binary_accuracy: 0.8869
Epoch 6/20
30/30 [==============================] - 1s 49ms/step - loss: 0.2101 - binary_accuracy: 0.9524 - val_loss: 0.3379 - val_binary_accuracy: 0.8850
Epoch 7/20
30/30 [==============================] - 1s 47ms/step - los

[0.2891996204853058, 0.8852400183677673]

Discussion: Using regularization was able to achieve a slightly better accuracy of 88.5%

5 b) Change: Adding Dropout

In [19]:
model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dropout(0.5))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dropout(0.5))
model.add(layers.Dense(1, activation='sigmoid'))


model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss='binary_crossentropy',
              metrics=['accuracy'])


model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy])


x_val = x_train[:10000]
partial_x_train = x_train[10000:]

y_val = y_train[:10000]
partial_y_train = y_train[10000:]


history = model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))

history_dict = history.history
history_dict.keys()


model = models.Sequential()
model.add(layers.Dense(16, activation='relu', input_shape=(10000,)))
model.add(layers.Dropout(0.5))
model.add(layers.Dense(16, activation='relu'))
model.add(layers.Dropout(0.5))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=4, batch_size=512)
results = model.evaluate(x_test, y_test)


results

Epoch 1/20
30/30 [==============================] - 2s 45ms/step - loss: 0.6374 - binary_accuracy: 0.6220 - val_loss: 0.4610 - val_binary_accuracy: 0.8497
Epoch 2/20
30/30 [==============================] - 1s 36ms/step - loss: 0.4707 - binary_accuracy: 0.8022 - val_loss: 0.3745 - val_binary_accuracy: 0.8590
Epoch 3/20
30/30 [==============================] - 1s 36ms/step - loss: 0.3934 - binary_accuracy: 0.8502 - val_loss: 0.3108 - val_binary_accuracy: 0.8835
Epoch 4/20
30/30 [==============================] - 1s 36ms/step - loss: 0.3355 - binary_accuracy: 0.8775 - val_loss: 0.2839 - val_binary_accuracy: 0.8914
Epoch 5/20
30/30 [==============================] - 1s 36ms/step - loss: 0.2886 - binary_accuracy: 0.8951 - val_loss: 0.2729 - val_binary_accuracy: 0.8928
Epoch 6/20
30/30 [==============================] - 1s 36ms/step - loss: 0.2465 - binary_accuracy: 0.9210 - val_loss: 0.2772 - val_binary_accuracy: 0.8906
Epoch 7/20
30/30 [==============================] - 1s 36ms/step - los

[0.276427686214447, 0.8890799880027771]

Discussion: Dropout too was able to increase the level of accuracy, this time to 88.9%